src/key/local.rs

```diff
@@ -0,0 +1,215 @@
+use std::fs::File;
+use std::io::Write;
+use std::path::PathBuf;
+
+use rcgen::KeyPair as RcKeyPair;
+use rcgen::{RsaKeySize, PKCS_RSA_SHA512};
+use thiserror::Error;
+
+use crate::key::creator::{CreationError, Creator, KeyPair, KeyType, Options, PublicKeyPem};
+
+/// Errors that can occur during local key creation.
+#[derive(Error, Debug)]
+pub enum LocalKeyCreationError {
+    /// Error that occurs when path to store the key is invalid.
+    #[error("invalid path: `{0}`")]
+    InvalidPath(String),
+    /// Error that occurs when key generation fails.
+    #[error("unable to generate key: `{0}`")]
+    UnableToGenerateKey(String),
+    /// Error that occurs when the private key file cannot be created.
+    #[error("unable to create private key file: `{0}`")]
+    UnableToCreatePrivateKeyFile(String),
+    /// Error that occurs when writing the private key fails.
+    #[error("unable to write private key: `{0}`")]
+    UnableToWritePrivateKey(String),
+}
+
+/// A creator for generating and managing cryptographic keys locally.
+#[derive(Debug)]
+pub struct LocalCreator {
+    /// The file path where the private key will be stored.
+    path: PathBuf,
+}
+
+impl Creator for LocalCreator {
+    /// Creates a cryptographic key based on the provided options and stores the private key locally.
+    ///
+    /// Returns the public key in PEM format, or an error if key creation fails.
+    fn create(&self, options: Options) -> Result<PublicKeyPem, CreationError> {
+        let key_pair = match options.key_type {
+            KeyType::Rsa4096 => rsa(options.key_type)?,
+        };
+
+        self.persist_private_key(options.name.as_str(), &key_pair.private_key)?;
+
+        Ok(key_pair.public_key)
+    }
+}
+
+impl TryFrom<PathBuf> for LocalCreator {
+    type Error = LocalKeyCreationError;
+
+    fn try_from(value: PathBuf) -> Result<Self, Self::Error> {
+        if !value.exists() {
+            return Err(LocalKeyCreationError::InvalidPath(String::from(
+                "local key path does not exist",
+            )));
+        }
+        if value.is_file() {
+            return Err(LocalKeyCreationError::InvalidPath(String::from(
+                "local key path needs to be a directory",
+            )));
+        }
+        Ok(LocalCreator { path: value })
+    }
+}
+
+impl LocalCreator {
+    /// Persists the private key to the specified file path.
+    fn persist_private_key(&self, name: &str, key: &[u8]) -> Result<(), LocalKeyCreationError> {
+        let mut file = File::create(&self.path.join(name).as_path())
+            .map_err(|e| LocalKeyCreationError::UnableToCreatePrivateKeyFile(e.to_string()))?;
+        file.write_all(key)
+            .map_err(|e| LocalKeyCreationError::UnableToWritePrivateKey(e.to_string()))?;
+        Ok(())
+    }
+}
+
+impl From<LocalKeyCreationError> for CreationError {
+    fn from(value: LocalKeyCreationError) -> Self {
+        CreationError::UnableToCreateKey(value.to_string())
+    }
+}
```

Comment:
I am a bit confused, I see the explicit map above, is this used?

Reply:
ohh, it is used for the rsa, could we make the translation in the code and add context? I think we agreed to reduce the #from and similar approaches

Reply:
is rsa, there is an explicit conversion now